### DISCUSSION TITLE
Produce finals in generated code

### DISCUSSION THREAD

**Discussion by H-Lo** [mapstruct/mapstruct#2601] [Ideas]
When you create simple mapping

    public abstract CountryDTO toDto(Country c);

generated code have no finals. Is there a way to get a finals in generated code?

    @Override
    public CountryDTO toDto(Country c) {
        if ( c == null ) {
            return null;
        }
        CountryDTO countryDTO = new CountryDTO();
        countryDTO.setId( c.getId() );
        countryDTO.setName( c.getName() );
        countryDTO.setStatus( c.getStatus() );
        return countryDTO;
    }
    
Desired code:
    
    @Override
    public CountryDTO toDto(   **final**   Country c) {
        if ( c == null ) {
            return null;
        }
        **final**   CountryDTO countryDTO = new CountryDTO();
        countryDTO.setId( c.getId() );
        countryDTO.setName( c.getName() );
        countryDTO.setStatus( c.getStatus() );
        return countryDTO;
    }

**Comment by filiphr** [comment]
Hey @H-Lo,

I see that you also created the issue #2602. If you want to start a discussion there is really no point in creating an issue as well.

The generated code you are suggesting is obsolete in modern Java, i.e. since Java 8, because they are effectively final.

Adding final will only make the code more difficult to read and will bring no real value

**Reply by H-Lo** [reply]
Well, this is partially true. Finals and effectively finals are not treated the same way with the compiler.
The code using finals usually has better performance.
Might seem silly to run for milliseconds, but for example when you have 200,000 documents to process generating two 2-paged PDF files for each, every millisecond you lose makes your processing last longer.
Few years ago I was testing finals on application mentioned above.
After applying finals I got 10% less time in processing which is pretty good as processing took few hours.

Simple class not using finals:

	public class WithoutFinals {
	    public static void main(String[] args) {
	        String one = "One";
	        String two = "Two";
	        String all = one + ' ' + two;
	        System.out.println(all);
	    }
	}

The same thing but using finals:

	public class WithFinals {
	    public static void main(final String[] args) {
	        final String one = "One";
	        final String two = "Two";
	        final String all = one + ' ' + two;
	        System.out.println(all);
	    }
	}

Running javac (1.8.0_181-b13) on both you get 657 bytes for the first one and 419 bytes for the second one which tells me there is pretty much difference while treating finals and effectively finals during compile time.

**Reply by H-Lo** [reply]
P.S. Sorry for creating an issue - the popup got me telling me "create an issue" and I thought that's the common practice as it told me to do so.

**Reply by filiphr** [reply]
You picked a really bad example 😉 . Look at the generated byte code between the two. The first one uses a String builder and the second one replaces the string it basically makes `String all = "One Two"`.

If we compare this with something that makes more sense e.g.

```
public class Within2Finals {
    public CountryDTO main(final Country c) {
        if ( c == null ) {
            return null;
        }
        final CountryDTO countryDTO = new CountryDTO();
        countryDTO.setId( c.getId() );
        countryDTO.setName( c.getName() );
        countryDTO.setStatus( c.getStatus() );
        return countryDTO;
    }
}
```

and 

```
public class WithoutFinals {
    public CountryDTO main(Country c) {
        if ( c == null ) {
            return null;
        }
        CountryDTO countryDTO = new CountryDTO();
        countryDTO.setId( c.getId() );
        countryDTO.setName( c.getName() );
        countryDTO.setStatus( c.getStatus() );
        return countryDTO;
    }
}
```

then the generated byte code for both is 707KB.

I've named one class `Within2Finals` in order for the class name to have same number of characters, otherwise the bytecode differs due to the fact that there are less characters in the class names.

If you would like to convince us that adding final is better then we would need to see some actual benchmarks such as JMH to see the impact that this would have. You can have a look at https://github.com/arey/java-object-mapper-benchmark which is an independent benchmark for different mappers. You could use it by copying the mapstruct generated mappers and adding final to them and using them.
